File: src/lib/validation.ts
export const validateName = (name: string): string | null => {
  const cleaned = name.trim();

  if (!cleaned) return "Name is required";
  if (cleaned.length < 2) return "Name must be at least 2 characters";
  if (cleaned.length > 100) return "Name must be less than 100 characters";

  if (!/^[\p{L}\p{N} .'\-/()]+$/u.test(cleaned)) {
    return "Name contains invalid characters";
  }

  return null;
};

export const validateAge = (age: number | null): string | null => {
  if (age === null || age === undefined) return "Age is required";
  if (!Number.isInteger(age)) return "Age must be a whole number";
  if (age < 0 || age > 150) return "Age must be between 0 and 150";
  return null;
};

export const validatePhone = (phone: string): string | null => {
  const cleaned = phone.trim();

  if (!cleaned) return null;
  if (cleaned.length < 10) return "Phone must be at least 10 characters";
  if (cleaned.length > 15) return "Phone must be less than 15 characters";

  if (!/^[0-9+\-() ]+$/.test(cleaned)) {
    return "Phone contains invalid characters";
  }

  return null;
};

export const validateGender = (gender: string): string | null => {
  if (!["Male", "Female", "Other"].includes(gender)) {
    return "Please select a valid gender";
  }

  return null;
};

export const validateAgeUnit = (unit: string): string | null => {
  if (!["Years", "Months", "Days"].includes(unit)) {
    return "Please select a valid age unit";
  }

  return null;
};

export const validateDoctorName = (name: string): string | null => {
  const cleaned = name.trim();

  if (!cleaned) return "Doctor name is required";
  if (cleaned.length < 2) return "Doctor name must be at least 2 characters";
  if (cleaned.length > 100)
    return "Doctor name must be less than 100 characters";

  return null;
};

export const validateAmount = (amount: number): string | null => {
  if (amount === null || amount === undefined) return "Amount is required";
  if (Number.isNaN(amount) || !Number.isFinite(amount))
    return "Amount must be valid";
  if (amount < 0) return "Amount cannot be negative";
  if (amount > 10_000_000) return "Amount exceeds maximum allowed value";

  const decimalPart = amount.toString().split(".")[1];

  if (decimalPart && decimalPart.length > 2) {
    return "Amount can have maximum 2 decimal places";
  }

  return null;
};

export const validateTestSelection = (testIds: number[]): string | null => {
  if (!testIds.length) return "Please select at least one test";
  if (testIds.length > 100) return "Too many tests selected";
  if (testIds.some((id) => id <= 0)) return "Invalid test selected";
  return null;
};

export const validateParameterValue = (value: string): string | null => {
  const cleaned = value.trim();

  if (!cleaned) return "Value is required";
  if (cleaned.length > 500) return "Value must be less than 500 characters";

  return null;
};

export const validatePatientForm = (
  name: string,
  age: number | null,
  ageUnit: string,
  gender: string,
  phone: string,
): Record<string, string> => {
  const errors: Record<string, string> = {};

  const nameError = validateName(name);
  if (nameError) errors.name = nameError;

  const ageError = validateAge(age);
  if (ageError) errors.age = ageError;

  const unitError = validateAgeUnit(ageUnit);
  if (unitError) errors.ageUnit = unitError;

  const genderError = validateGender(gender);
  if (genderError) errors.gender = genderError;

  const phoneError = validatePhone(phone);
  if (phoneError) errors.phone = phoneError;

  return errors;
};

export const hasErrors = (errors: Record<string, string>): boolean => {
  return Object.values(errors).some(Boolean);
};

export const getFirstError = (
  errors: Record<string, string>,
): string | null => {
  const firstKey = Object.keys(errors).find((key) => errors[key]);
  return firstKey ? errors[firstKey] : null;
};

File: src-tauri/src/utils/validation.rs
/// ✅ Input Validation Module
/// Validates all user inputs before database operations
/// Prevents SQL injection, invalid data, and business logic violations

use crate::errors::{AppError, AppResult};


/// Validate patient name
/// - Not empty
/// - Not too long (max 100 chars)
/// - No special SQL characters
pub fn validate_patient_name(name: &str) -> AppResult<()> {
    if name.trim().is_empty() {
        return Err(AppError::ValidationError(
            "Patient name cannot be empty".to_string(),
        ));
    }

    if name.len() > 100 {
        return Err(AppError::ValidationError(
            "Patient name too long (max 100 characters)".to_string(),
        ));
    }

    // ⚠️ Allow basic letters, spaces, hyphens, apostrophes
    if !name.chars().all(|c| c.is_alphabetic() || " -'".contains(c)) {
        return Err(AppError::ValidationError(
            "Patient name contains invalid characters".to_string(),
        ));
    }

    Ok(())
}

/// Validate patient age
pub fn validate_age(age: i32) -> AppResult<()> {
    if age < 0 || age > 150 {
        return Err(AppError::ValidationError(
            "Age must be between 0 and 150".to_string(),
        ));
    }
    Ok(())
}

/// Validate phone number (basic validation)
pub fn validate_phone(phone: &str) -> AppResult<()> {
    if phone.is_empty() {
        return Ok(()); // Optional field
    }

    if phone.len() < 10 || phone.len() > 15 {
        return Err(AppError::ValidationError(
            "Phone number must be 10-15 digits".to_string(),
        ));
    }

    if !phone.chars().all(|c| c.is_numeric() || "+-() ".contains(c)) {
        return Err(AppError::ValidationError(
            "Phone contains invalid characters".to_string(),
        ));
    }

    Ok(())
}

/// Validate amount (price, payment)
pub fn validate_amount(amount: f64) -> AppResult<()> {
    if amount < 0.0 {
        return Err(AppError::ValidationError(
            "Amount cannot be negative".to_string(),
        ));
    }

    if amount.is_nan() || amount.is_infinite() {
        return Err(AppError::ValidationError(
            "Invalid amount value".to_string(),
        ));
    }

    // Max amount: 10 million (reasonable upper bound)
    if amount > 10_000_000.0 {
        return Err(AppError::ValidationError(
            "Amount exceeds maximum allowed value".to_string(),
        ));
    }

    Ok(())
}

/// Validate doctor name
pub fn validate_doctor_name(name: &str) -> AppResult<()> {
    if name.trim().is_empty() {
        return Err(AppError::ValidationError(
            "Doctor name cannot be empty".to_string(),
        ));
    }

    if name.len() > 100 {
        return Err(AppError::ValidationError(
            "Doctor name too long (max 100 characters)".to_string(),
        ));
    }

    Ok(())
}

/// Validate test parameter value
pub fn validate_parameter_value(value: &str) -> AppResult<()> {
    if value.trim().is_empty() {
        return Err(AppError::ValidationError(
            "Parameter value cannot be empty".to_string(),
        ));
    }

    if value.len() > 500 {
        return Err(AppError::ValidationError(
            "Parameter value too long (max 500 characters)".to_string(),
        ));
    }

    Ok(())
}

/// Validate multiple test IDs
pub fn validate_test_ids(test_ids: &[i32]) -> AppResult<()> {
    if test_ids.is_empty() {
        return Err(AppError::ValidationError(
            "At least one test must be selected".to_string(),
        ));
    }

    if test_ids.len() > 100 {
        return Err(AppError::ValidationError(
            "Too many tests selected (max 100)".to_string(),
        ));
    }

    // Check for negative IDs
    if test_ids.iter().any(|&id| id <= 0) {
        return Err(AppError::ValidationError(
            "Invalid test ID provided".to_string(),
        ));
    }

    Ok(())
}

/// Validate positive integer (for IDs)
pub fn validate_positive_id(id: i32, field_name: &str) -> AppResult<()> {
    if id <= 0 {
        return Err(AppError::ValidationError(format!(
            "{} must be a positive number",
            field_name
        )));
    }
    Ok(())
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_validate_name_valid() {
        assert!(validate_patient_name("John Doe").is_ok());
        assert!(validate_patient_name("Mary-Jane O'Connor").is_ok());
    }

    #[test]
    fn test_validate_name_empty() {
        assert!(validate_patient_name("").is_err());
        assert!(validate_patient_name("   ").is_err());
    }

    #[test]
    fn test_validate_age() {
        assert!(validate_age(25).is_ok());
        assert!(validate_age(-1).is_err());
        assert!(validate_age(200).is_err());
    }

    #[test]
    fn test_validate_amount() {
        assert!(validate_amount(100.50).is_ok());
        assert!(validate_amount(-50.0).is_err());
        assert!(validate_amount(20_000_000.0).is_err());
    }
}
